### ISSUE TITLE
Is this N/3 correct?

### ISSUE THREAD

**Issue by alecjacobson** [EnzymeAD/Enzyme#1886]
https://github.com/EnzymeAD/Enzyme/blob/2188ad8f1401886eca1e2045abad6b60424fc2c5/enzyme/test/Integration/Sparse/ringspring3Drestlengthone.cpp#L24

`N` seems to be the number of 3D nodes. So that `pos` and `x` are arrays of length `3*N`.

This `N/3` seems to mean that only first ⅓ of the springs are considered.

**Comment by alecjacobson** [comment]
And if it's changed to `N` then it seems the `i*3+4` etc. need a `%(N*3)` to wrap around the circle of springs.

**Comment by wsmoses** [comment]
Yeah I think I agree here. Fortunately this is just a test for correct
compilation and not a benchmark.

**Comment by wsmoses** [comment]
I'm not sure about the % . I need to double check what this test is validating for, but the wrap around in the actual benchmarks is done via the todense to create the actual double pointers (and accessing them will implicitly wrap around)

**Comment by alecjacobson** [comment]
Thanks. @martinjm97 do you mind adding me as a collaborator on that repo so I can copy the benchmark for our comparison?

**Comment by martinjm97** [comment]
Sorry, I just saw this. I've added you to the repo.

**Comment by martinjm97** [comment]
The benchmarking script is here:

https://github.com/dromaiidae/sparse_deriv/blob/siggraph_benchmarking/benchmark/SiggraphPerformance/ring_spring_rest_length_one_timing.py

The code that we run is in the test is here:
https://github.com/dromaiidae/sparse_deriv/blob/dc973f23d672232089c30efe9bccabf48ebab765/benchmark/SiggraphPerformance/ring_spring_rest_length_one_timing.py#L56

**Comment by wsmoses** [comment]
I don't think I have access to that repo, @deomaiidae if you can add me

**Comment by martinjm97** [comment]
@alecjacobson the code has a bug as you pointed out. Moreover, the benchmarking script calls the file at this location. However, when I ran the benchmarking script, I ran a version of the code that fixed this bug (and also printed out how much time the benchmark took to run). I've made a PR of the code I ran: https://github.com/EnzymeAD/Enzyme/pull/1889.

**Comment by wsmoses** [comment]
Yeah that post sparse todense is the one I remember working with you on during benchmarking, so that looks right.
